library/src/com/google/maps/android/kml/KmlRenderer.java

```diff
@@ -0,0 +1,870 @@
+     *
+     * @param groundOverlayUrl url of ground overlay
+     * @param groundOverlays   hashmap of ground overlays to add to the map
+     */
+    private void addGroundOverlayToMap(String groundOverlayUrl,
+            HashMap<KmlGroundOverlay, GroundOverlay> groundOverlays) {
+        BitmapDescriptor groundOverlayBitmap = BitmapDescriptorFactory
+                .fromBitmap(mImagesCache.get(groundOverlayUrl));
+        for (KmlGroundOverlay groundOverlay : groundOverlays.keySet()) {
+            if (groundOverlay.getImageUrl().equals(groundOverlayUrl)) {
+                GroundOverlayOptions groundOverlayOptions = groundOverlay.getGroundOverlayOptions()
+                        .image(groundOverlayBitmap);
+                groundOverlays.put(groundOverlay, mMap.addGroundOverlay(groundOverlayOptions));
+            }
+        }
+    }
+
+    /**
+     * Adds ground overlays in containers from a given URL onto the map
+     *
+     * @param groundOverlayUrl url of ground overlay
+     * @param kmlContainers    containers containing ground overlays to add to the map
+     */
+    private void addGroundOverlayInContainerGroups(String groundOverlayUrl,
+            Iterable<KmlContainer> kmlContainers) {
+        for (KmlContainer container : kmlContainers) {
+            addGroundOverlayToMap(groundOverlayUrl, container.getGroundOverlayHashMap());
+            if (container.hasNestedKmlContainers()) {
+                addGroundOverlayInContainerGroups(groundOverlayUrl,
+                        container.getNestedKmlContainers());
+            }
+        }
+    }
+
+    /**
+     * Downloads images for use as marker icons
+     */
+    private class MarkerIconImageDownload extends AsyncTask<String, Void, Bitmap> {
+
+        private final String mIconUrl;
+
+        /**
+         * Creates a new IconImageDownload object
+         *
+         * @param iconUrl URL of the marker icon to download
+         */
+        public MarkerIconImageDownload(String iconUrl) {
+            mIconUrl = iconUrl;
+        }
+
+        /**
+         * Downloads the marker icon in another thread
+         *
+         * @param params String varargs not used
+         * @return Bitmap object downloaded
+         */
+        @Override
+        protected Bitmap doInBackground(String... params) {
+            try {
+                return BitmapFactory.decodeStream((InputStream) new URL(mIconUrl).getContent());
+            } catch (MalformedURLException e) {
+                e.printStackTrace();
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+            return null;
+        }
+
+        /**
+         * Adds the bitmap to the cache and adds the bitmap to the markers
+         *
+         * @param bitmap bitmap downloaded
+         */
+        @Override
+        protected void onPostExecute(Bitmap bitmap) {
+            if (bitmap == null) {
+                throw new NullPointerException("Image not found!");
```

Comment:
This seems excessive.  This might be because we're offline or the server is down and crashing is probably not the right behavior.